File: Jump_to_Algorithm3/week13/SoohyunKim/2138_전구와스위치.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

function solution(input) {
  const N = +input[0];
  const current = input[1].split('').map(Number); 
  const target = input[2].split('').map(Number); 

  // 스위치 눌러 전구 상태를 바꾸는 함수
  const reverse = (arr, index) => {
    for (let i = index - 1; i <= index + 1; i++) {
      if (i >= 0 && i < N) {
        arr[i] = 1 - arr[i];
      }
    }
  };

  // 최소 스위치 횟수를 계산
  const countSwitches = (initialState, pressFirst) => {
    const state = [...initialState];
    let count = 0;

    // 첫 번째 스위치를 누르는 경우
    if (pressFirst) {
      reverse(state, 0);
      count++;
    }

    // 두 번째 전구부터 끝까지 처리
    for (let i = 1; i < N; i++) {
      if (state[i - 1] !== target[i - 1]) {
        reverse(state, i);
        count++;
      }
    }

    // 목표 상태와 일치 여부 확인
    return state.every((e, i) => e === target[i]) ? count : -1;
  };

  // 0번 스위치를 누르지 않은 경우와 누른 경우 두가지 계산
  const firstOff = countSwitches(current, false);
  const firstOn = countSwitches(current, true);

  // 최소값 계산
  if (firstOff === -1 && firstOn === -1) {
    return -1;
  } else if (firstOff === -1) {
    return firstOn;
  } else if (firstOn === -1) {
    return firstOff;
  } else {
    return Math.min(firstOff, firstOn);
  }
}

console.log(solution(input));

File: Jump_to_Algorithm3/week13/HyeonJiLee/2138 전구와 스위치.py
N = int(input())
current = input()
target = input()

#첫 번째 스위치를 누른 경우와 누르지 않은 경우...를 나누어 비교...

def onoff(arr, i):
    return '1' if arr[i] == '0' else '0'

def swtich(state, count):
    for i in range(1, N):
        #두번째 스위치부터 탐색
        if state[i-1] != target[i-1]:
            count += 1
            #스위치 켜고끄기(i-1, i)
            state = state[:i-1] + onoff(state, i-1) + onoff(state, i) + state[i+1:]
            #맨 마지막 스위치가 아닌 경우 i+1번째도 바꾸기
            if i < N-1:
                state = state[:i+1] + onoff(state, i+1) + state[i+2:]
    return count if state == target else 100001 #누를 수 있는 최대 횟수 + 1

# 첫 번째 스위치를 누르지 않은 경우
result1 = swtich(current, 0)

# 첫 번째 스위치를 누른 경우
flipped = onoff(current, 0) + onoff(current, 1) + current[2:]
result2 = swtich(flipped, 1)

min_switches = min(result1, result2)
min_switches = min_switches if min_switches != 100001 else -1

print(min_switches)